packages/theme-language-server-common/src/json/completions/providers/BlockTypeCompletionProvider.ts

```diff
@@ -0,0 +1,107 @@
+import {
+  AppBlockSchema,
+  deepGet,
+  isError,
+  SectionSchema,
+  ThemeBlockSchema,
+  ThemeSchemaType,
+} from '@shopify/theme-check-common';
+import { JSONPath } from 'vscode-json-languageservice';
+import { JSONCompletionItem } from 'vscode-json-languageservice/lib/umd/jsonContributions';
+import { CompletionItemKind } from 'vscode-languageserver-protocol';
+import { isLiquidRequestContext, RequestContext } from '../../RequestContext';
+import { fileMatch } from '../../utils';
+import { JSONCompletionProvider } from '../JSONCompletionProvider';
+
+export type GetThemeBlockNames = (uri: string) => Promise<string[]>;
+
+/**
+ * The BlockTypeCompletionProvider offers value completions of the
+ * `blocks.[].type` property inside section and theme block `{% schema %}` tags.
+ *
+ * @example
+ * {% schema %}
+ * {
+ *   "blocks": [
+ *     { "type": "█" },
+ *   ]
+ * }
+ * {% endschema %}
+ */
+export class BlockTypeCompletionProvider implements JSONCompletionProvider {
+  private uriPatterns = [/^.*\/(sections|blocks)\/[^\/]*\.liquid$/];
+
+  constructor(private getThemeBlockNames: GetThemeBlockNames) {}
+
+  async completeValue(context: RequestContext, path: JSONPath): Promise<JSONCompletionItem[]> {
+    if (
+      !fileMatch(context.doc.uri, this.uriPatterns) ||
+      !isLiquidRequestContext(context) ||
+      !isBlockDefinitionPath(path)
+    ) {
+      return [];
+    }
+    const { doc } = context;
+    const schema = await doc.getSchema();
+
+    // Can't complete if we can't parse the schema
+    if (!schema || isError(schema.parsed) || !isSectionOrBlockSchema(schema)) {
+      return [];
+    }
+
+    // Local blocks have their type defined in the schema, there's nothing to complete
+    if (hasLocalBlockDefinitions(schema)) return [];
+
+    const blockNames = await this.getThemeBlockNames(doc.uri);
+
+    return blockNames.map((name) => ({
+      kind: CompletionItemKind.Value,
+      label: `"${name}"`,
+      insertText: `"${name}"`,
+    }));
+  }
+}
+
+export function isBlockDefinitionPath(path: JSONPath) {
+  return path.at(0) === 'blocks';
+}
+
+export function isPresetBlockPath(path: JSONPath) {
+  return path.at(0) === 'presets';
+}
+
+export function isBlockTypePath(path: JSONPath) {
+  // We have these cases to support:
+  // - top level blocks.[].type
+  // - presets.[](recursive .blocks.[].type)
+  // - presets.[](recursive .blocks.{}.type)
+  const topLevel = path.at(0);
+  if (topLevel !== 'blocks' && topLevel !== 'presets') return false;
+  if (path.length < 4) return false; // minimum path length
+  const [shouldBeBlocks, _idOrIndex, shouldBeType] = path.slice(-3);
+  return shouldBeBlocks === 'blocks' && shouldBeType === 'type';
```

Comment:
Same that was something I used before I planned to split #628 to a different provider